### PR TITLE
S3ArtifactDriver added

### PR DESCRIPTION
**Pull request by Raalsky**

### CODE REVIEW

File: neptune/new/internal/artifacts/s3.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            for remote_object in remote_storage.objects.filter(Prefix=prefix):
+                remote_key = pathlib.Path(name or '') / pathlib.Path(remote_object.key[len(prefix):])
+
+                stored_files.append(
+                    ArtifactFileData(
+                        file_path=str(remote_key),
+                        file_hash=remote_object.e_tag.strip('"'),
+                        type=ArtifactFileType.S3.value,
+                        metadata={
+                            "location": f's3://{bucket_name}{remote_object.key}',
```
Comment (pkasprzyk) [inline neptune/new/internal/artifacts/s3.py:59]: are you 100% sure that `remote_object.key` always `startswith('/')`? if not, please accomodate (lstrip + prepend)

File: neptune/new/internal/artifacts/s3.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        location = file_definition.metadata.get('location')
+        url = urlparse(location)
+        bucket_name, path = url.netloc, url.path
+
+        remote_storage = boto3.resource('s3')
+
+        try:
+            # pylint: disable=no-member
+            bucket = remote_storage.Bucket(bucket_name)
+            bucket.download_file(path, str(destination))
```
Comment (pkasprzyk) [inline neptune/new/internal/artifacts/s3.py:84]: I'd expect `bucket.download_file(path, str(destination / file_definition.file_path))` here?

File: neptune/new/internal/artifacts/s3.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        except NoCredentialsError:
+            raise NeptuneRemoteStorageCredentialsException()
+        except (remote_storage.meta.client.exceptions.NoSuchBucket,
```
Comment (pkasprzyk) [inline neptune/new/internal/artifacts/s3.py:87]: again duplicate exception?

**Review by pkasprzyk — CHANGES_REQUESTED** [review]

File: tests/neptune/new/internal/artifacts/test_s3.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            self.assertEqual('2f249230a8e7c2bf6005ccd2679259ec', md5(local_destination))
+
+    def test_single_retrieval(self):
+        files = S3ArtifactDriver.get_tracked_files(f"s3://{self.bucket_name}/path/to/file1")
+
+        self.assertEqual(1, len(files))
+        self.assertIsInstance(files[0], ArtifactFileData)
+        self.assertEqual(ArtifactFileType.S3.value, files[0].type)
+        self.assertEqual('2f249230a8e7c2bf6005ccd2679259ec', files[0].file_hash)
+        self.assertEqual('.', files[0].file_path)
```
Comment (Raalsky) [inline tests/neptune/new/internal/artifacts/test_s3.py:79]: @pkasprzyk Are we sure how do we should handle this?

File: neptune/new/internal/artifacts/s3.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                        metadata={
+                            "location": f's3://{bucket_name}{remote_object.key}',
+                            "last_modified": remote_object.last_modified,
+                            "file_size": remote_object.size
+                        }
+                    )
+                )
+        except NoCredentialsError:
+            raise NeptuneRemoteStorageCredentialsException()
+        except (remote_storage.meta.client.exceptions.NoSuchBucket,
```
[the remark this answers is not in the document]
Comment (Raalsky) [reply neptune/new/internal/artifacts/s3.py:67]: XD. Thanks, just a typo when c&p :wink:

File: neptune/new/internal/artifacts/s3.py (hunk 2 above)
Comment (Raalsky) [reply neptune/new/internal/artifacts/s3.py:84]: Ok, It wasn't specified :wink:

File: tests/neptune/new/internal/artifacts/test_s3.py (hunk 4 above)
Comment (pkasprzyk) [reply tests/neptune/new/internal/artifacts/test_s3.py:79]: missed that, thanks -- "we're" quite sure it should be `self.assertEqual('file1', files[0].file_path)`

we should treat differently:
1. a call to `get_tracked_files()` with prefix pointing to an existing object (then name of this object becomes file name) 2. the other situaction (then all objects with this prefix should be treated as in the current code, i.e. stripped from prefix)

File: tests/neptune/new/internal/artifacts/test_s3.py [hunk 6]
```diff
@@ -60,14 +60,17 @@ def test_file_download(self):
...
 
             S3ArtifactDriver.download_file(
                 destination=local_destination,
                 file_definition=artifact_file
             )
 
-            self.assertEqual('2f249230a8e7c2bf6005ccd2679259ec', md5(local_destination))
+            import os
+
+            print(os.listdir('.'))
```
Comment (pkasprzyk) [inline tests/neptune/new/internal/artifacts/test_s3.py:72]: ahem

**Pull request merged**